atspi-macros/src/lib.rs

```diff
@@ -0,0 +1,192 @@
+                };
+
+                let tname = std::any::type_name::<Self>().to_string();
+                let member = tname.strip_suffix("Event").unwrap();
+                let error = format!("specific type's member: {} != msg type member: {:?}", member, msg_member);
+                Err(crate::AtspiError::MemberMatch(error))
+            }
+        }
+
+        impl<'a> Signified for #name {
+            type Inner = AtspiEvent;
+            fn inner(&self) -> &Self::Inner {
+                &self.0
+            }
+
+            fn properties(&self) -> &HashMap<String, OwnedValue> {
+                self.0.properties()
+            }
+        }
+    };
+
+    // Return the expanded code as a token stream
+    TokenStream::from(expanded)
+}
+
+//#[proc_macro_derive(TryFromMessage)]
+#[proc_macro_attribute]
+pub fn try_from_zbus_message(attr: TokenStream, input: TokenStream) -> TokenStream {
+    let item_struct = parse_macro_input!(input as ItemStruct);
+    // Parse the input token stream into a syntax tree
+    let name = item_struct.ident.clone();
+
+    // Remove the suffix "Event" from the name of the struct
+    let name_string = name.to_string();
+    let member = name_string.strip_suffix("Event").unwrap();
+
+    let args_parsed: Vec<FromZbusMessageParam> = parse_macro_input!(attr as AttributeArgs)
+        .into_iter()
+        .filter_map(|nm| match nm {
+            // Only select certain tokens
+            NestedMeta::Meta(Meta::NameValue(MetaNameValue {
+                path,
+                eq_token: _,
+                lit: Lit::Str(lstr),
+            })) => Some(
+                // Convert the sigment of the path to a string
+                (
+                    path.segments
+                        .into_iter()
+                        .map(|seg| seg.ident.to_string())
+                        .collect::<Vec<String>>()
+                        .swap_remove(0),
+                    // get the raw value of the LitStr
+                    lstr.value(),
+                ),
+            ),
+            _ => None,
+        })
+        // convert the (String, LitStr) tuple to a custom type which only accepts certain key/value pairs
+        .map(|(k, v)| FromZbusMessageParam::from(k, v))
+        .collect();
+
+    let body_type = match args_parsed
+        .get(0)
+        .expect("There must be exactly one argument to the macro.")
+    {
+        FromZbusMessageParam::Body(body_type) => body_type,
+        _ => panic!("The body parameter must be a type."),
+    };
+
+    // Generate the expanded code
+    let expanded = quote! {
+                #item_struct
+        impl TryFrom<Arc<Message>> for #name {
+            type Error = AtspiError;
+
+            fn try_from(message: Arc<Message>) -> Result<Self, Self::Error> {
+                                if message.member() != Some(MemberName::from_static_str(#member)?) {
+                                    return Err(AtspiError::CacheVariantMismatch);
```

Comment:
We likely want a more generic `Error` variant here.
Along these lines perhaps?: `Error::MemberMatch(#name.to_string())`

Reply:
Yeah this would be a MemberMismatchError, I suppose?

Reply:
Sounds good to me!